src/vsr/replica.zig

```diff
@@ -3719,6 +3719,57 @@ pub fn ReplicaType(
             return self.op_checkpoint_next() + constants.lsm_batch_multiple;
         }
 
+        /// Returns the oldest op that the replica must/(is permitted to) repair.
+        /// The goal is to repair as much as possible without triggering unnecessary state transfer.
+        ///
+        /// - Repairing uncommitted ops is necessary as primary — we will need them to proceed.
+        /// - Repairing committed + not-yet-checkpointed ops is useful because we might crash,
+        ///   in which case we will need them to recover.
+        /// - Repairing committed + checkpointed ops:
+        ///   - backups don't repair checkpointed ops, since there is no guarantee that they will
+        ///     ever be available (if our head is behind), and we don't want to stall the new view
+        ///     startup.
+        ///   - primaries do repair checkpointed ops — as many as are guaranteed to exist anywhere in
+        ///     the cluster, so that they can help lagging backups catch up.
```

Comment:
Hm, I don't understand this bit: why primary can repair beyond checkpoint, but a backup cant. :thinking: or maybe I do?

If we are a backup, we don't know the cluster's state. The cluster might have moved ahead of us, so preparess before checkpoint might not be available.

If we are the primary, we know the cluster's state! The only way cluster can move ahead is if there's a view change (but then we'll view jump and no longer be primary).

So, basically, when the primary repairs, it is guaranteed to either repair or view jump.

If a backup repairs, in general it _might_ get stuck if prepare it requests is no longer available. This might happen even if we request very few prepares, but that's OK, as we'll switch to state transfer if repair gets stuck.

In other words, primary's repair always suceeds, so primary tries to request **as much as we could** (so, journal-full)

Backup repair might fail, so we try to request as **little as we must**

Reply:
> since there is no guarantee that they will ever be available

Hm, this is not the whole story, right? This is not only about availability, but also about correcness, as per

```
// It is critical that we do not repair checkpointed ops; their slots now belong
```

?

Reply:
> If we are a backup, we don't know the cluster's state. The cluster might have moved ahead of us, so preparess before checkpoint might not be available.
> If we are the primary, we know the cluster's state! The only way cluster can move ahead is if there's a view change (but then we'll view jump and no longer be primary).

This is exactly right! The backups know the _past_ state of the primary, but if the primary has advanced since then, they might be requesting ops that are unavailable. The primary knows its state right now.

> Hm, this is not the whole story, right? This is not only about availability, but also about correcness, as per

That's a good point, I'll update the comment. If backups repaired before the checkpoint they could overwrite slots which belong to the current wrap, which could lead to the backup incorrectly nacking a message.